### ISSUE TITLE
Ratings support?

### ISSUE THREAD

**Issue by sardaukar** [sentriz/gonic#31]
Is implementing ratings per track a goal?

**Comment by sentriz** [comment]
yes I would hope to do this in the future. should not be too hard. just need some time

**Comment by davidhampgonsalves** [comment]
@sentriz do you have ideas for how you plan on implementing this? I see that nicksellen/audiotags only supports read operations through taglib. Were you thinking of adding the ability to change ratings to that first?

**Comment by sentriz** [comment]
You mean to say the ratings are stored in the audio tags themselves?

**Comment by sentriz** [comment]
My idea was a track rating table instead of tags. With columns track_id, user_id, rating - to make sure ratings are per user. But I'd like to know what airsonic etc does

**Comment by davidhampgonsalves** [comment]
Looks like that's what airsonic is doing as well:

https://github.com/airsonic/airsonic/blob/828c1b2674b1b5bf4f96454ed5c154a0c074f3bb/airsonic-main/src/main/java/org/airsonic/player/controller/SetRatingController.java

I need the files to be the source of truth so I'll do something different. 
Thanks

**Comment by sentriz** [comment]
if you follow that path:
https://github.com/airsonic/airsonic/blob/828c1b2674b1b5bf4f96454ed5c154a0c074f3bb/airsonic-main/src/main/java/org/airsonic/player/controller/SetRatingController.java#L62
-> 
https://github.com/airsonic/airsonic/blob/828c1b2674b1b5bf4f96454ed5c154a0c074f3bb/airsonic-main/src/main/java/org/airsonic/player/service/RatingService.java#L76
->
https://github.com/airsonic/airsonic/blob/828c1b2674b1b5bf4f96454ed5c154a0c074f3bb/airsonic-main/src/main/java/org/airsonic/player/dao/RatingDao.java#L80

they seem to be doing the same as I suggested

**Comment by davidhampgonsalves** [comment]
Yes, sorry I was unclear, I meant that what you suggested is what airsonic does.

**Comment by sentriz** [comment]
ohh sorry yes. I misread :) if I did end up implementing ratings the way airsonic does. you could probably make a little script to update the gonic db's ratings for a given user and list of tracks

**Comment by davidhampgonsalves** [comment]
Yea I think I would do that. Also thanks for exposing gonicscan, really handy.

**Comment by sentriz** [comment]
also, looking at http://www.subsonic.org/pages/api.jsp "setRating"
this depends on #41 
as pointed out: https://github.com/sentriz/gonic/issues/41#issuecomment-593503130
since the rating id can be a track, album, etc. and gonic uses primary keys for it's track, album, etc ids

**Comment by r0bbie** [comment]
Would supporting favouriting songs come under this, or be a separate feature request?

**Comment by sentriz** [comment]
just  merged the ambig ids branch. this feature should now be possible. eg handler

```go
func (c *Controller) ServeSetRating(r *http.Request) *spec.Response {
    params := r.Context().Value(CtxParams).(params.Params)
    id, err := params.GetID("id")
    // handle err
    switch id.Type {
    case spec.Artist:
        // rate artist
    case spec.Album:
        // rate album
    case spec.Track:
        // rate track
    }
}
```

hey @r0bbie they're not exactly the same of but very similar to implement (mostly copy paste `>:)`)

**Comment by sentriz** [comment]
closing in favour of #171

**Comment by sentriz** [comment]
ratings and stars are now shipped in master / :nightly . thanks to @brian-doherty 

next release hopefully soon:
https://github.com/sentriz/gonic/milestone/1
